### PR TITLE
Integrate Lightbeam with latest Wasmtime master

### PR DESCRIPTION
**Pull request by Vurich**
This PR closes #979, as it integrates those changes along with the removal of `multi_mut` mentioned in that PR. All the spec tests are passing now, there were many weird bugs related to Lightbeam's lack of proper abstraction around the actual codegen, so hopefully this kind of change won't be so difficult in the future. This PR also implements trap info and the `ModuleAddressMap` debuginfo tagging in Lightbeam, although it's hard to tell if it is integrated correctly as there are no tests for it and as far as I can tell it's not possible to dump this information anywhere.

You'll see that because of the aforementioned lack of abstraction, a lot of the changes for the caller vmctx work seem pretty janky. I have plans to remove all the jankyness, but I figured it was most important to just get it working again before starting on any work doing rearchitecturing.

Additionally, there are a lot of indentation changes in `function_body` which make the +/- inaccurate and the diff hard to read, so consider using the "Hide whitespace changes" option in the diff settings.

### CODE REVIEW

File: crates/lightbeam/src/backend.rs [hunk 1]
```diff
@@ -1874,12 +1904,12 @@ macro_rules! load {
...
                     dynasm!(ctx.asm
-                        ; cmp [
+                        ; cmp Rq(addr_reg.rq().unwrap()), [
                             Rq(reg.unwrap_or(vmctx).rq().unwrap()) +
                                 mem_offset +
                                 ctx.module_context.vmmemory_definition_current_length() as i32
-                        ], Rq(addr_reg.rq().unwrap())
-                        ; jna =>trap_label.0
+                        ]
+                        ;; ctx.trap_if(cc::GE_U, TrapCode::HeapOutOfBounds)
```
Comment (0x7CFE) [inline crates/lightbeam/src/backend.rs]: Redundant `;`?

File: crates/lightbeam/src/microwasm.rs [hunk 2]
```diff
@@ -1196,47 +1362,53 @@ where
...
             WasmOperator::F32Const { .. } => sig!(() -> (F32)),
             WasmOperator::F64Const { .. } => sig!(() -> (F64)),
 
-            WasmOperator::RefNull => {
-                return Err(Error::Microwasm("RefNull unimplemented".to_string()))
-            }
-            WasmOperator::RefIsNull => {
-                return Err(Error::Microwasm("RefIsNull unimplemented".to_string()))
-            }
+            // WasmOperator::RefNull => {
```
Comment (0x7CFE) [inline crates/lightbeam/src/microwasm.rs]: Was this commented out intentionally?

File: crates/lightbeam/Cargo.toml [hunk 3]
```diff
@@ -11,19 +11,24 @@ keywords = ["webassembly", "wasm", "compile", "compiler", "jit"]
...
+auto_enums = "0.7"
+capstone = "0.6.0"
+chain-one = "0.1"
+cranelift-codegen = { path = "../../cranelift/codegen", version = "0.59.0" }
+derive_more = "0.99"
 dynasm = "0.5.2"
 dynasmrt = "0.5.2"
-wasmparser = "0.51.2"
-memoffset = "0.5.3"
+hashers = "1.0"
```
Comment (sunfishcode) [inline crates/lightbeam/Cargo.toml]: Is this dependency on `hashers` needed?

File: crates/lightbeam/Cargo.toml [hunk 4]
```diff
@@ -11,19 +11,24 @@ keywords = ["webassembly", "wasm", "compile", "compiler", "jit"]
...
+smallvec = "1.0.0"
+staticvec = "0.8"
 thiserror = "1.0.9"
-cranelift-codegen = { path = "../../cranelift/codegen", version = "0.59.0" }
-multi_mut = "0.1"
-either = "1.5"
 typemap = "0.3"
-more-asserts = "0.2.1"
+wasmparser = "0.51.2"
+wasm-reader = "0.2"
```
Comment (sunfishcode) [inline crates/lightbeam/Cargo.toml]: Switching the wasm decoder is a very big change, and it's not clear what motivates it, or how it's connected to the rest of the PR. Could you split this part into a separate PR so that we can discuss it separately?

File: crates/lightbeam/Cargo.toml [hunk 5]
```diff
@@ -11,19 +11,24 @@ keywords = ["webassembly", "wasm", "compile", "compiler", "jit"]
 edition = "2018"
 
 [dependencies]
-smallvec = "1.0.0"
+auto_enums = "0.7"
+capstone = "0.6.0"
+chain-one = "0.1"
```
Comment (sunfishcode) [inline crates/lightbeam/Cargo.toml]: Would it be possible to split out the changes that introduce the dependency on `chain_one` into a separate PR? Changes introducing new external dependencies are something we'd like to consider carefully, and it's difficult to do that when it's bundled together with the rest of the changes here.

File: crates/lightbeam/Cargo.toml (hunk 4 above)
Comment (Vurich) [reply crates/lightbeam/Cargo.toml]: So as we weren't integrated with the latest Wasmtime for a while, this PR includes some changes that were implemented separately, before the integration. This work was part of two features: firstly, implementing trap information, and secondly converting Lightbeam to use `R: Read` everywhere instead of requiring a slice. It was used for the trap information as it hands control of the reader back to Lightbeam, so that Lightbeam can act independently of the parser WRT source location and other metadata. Specifically, it makes it so that Lightbeam can get the current position without requiring `wasmparser` to implement a specific accessor for the specific property we want. Both of these features would be very difficult to include separately to the rest of the work, as the integration with master was implemented on top of them. You can check the tests for `wasm-reader`, though, as it has an extensive test suite comparing it against `wasmparser`.

File: crates/environ/src/func_environ.rs [hunk 6]
```diff
@@ -494,6 +499,11 @@ impl lightbeam::ModuleContext for FuncEnvironment<'_> {
             .map(DefinedMemoryIndex::as_u32)
     }
 
+    fn vmctx_builtin_function(&self, func_index: u32) -> u32 {
+        self.offsets
+            .vmctx_builtin_function(BuiltinFunctionIndex::from_u32(func_index))
```
[the remark this answers is not in the document]
Comment (Vurich) [reply crates/environ/src/func_environ.rs]: Unfortunately the `BuiltinFunctionIndex` passed to this function and the `BuiltinFunctionIndex` used in the implementation of this function are not the same. I had to copy-paste the implementation to get the correct indices for the built-in functions and use `u32` for the types in the trait because otherwise I get cyclical dependencies - many of the types I would need to use are defined in `wasmtime-environ`, but Lightbeam is itself a dependency of `wasmtime-environ`.

[the remark this answers is not in the document]
Comment (sunfishcode) [reply crates/environ/src/func_environ.rs:572]: Ok. Could you add TODO comments mentioning that, as we discussed offline, the better approach here would be to start factoring out some of this into a separate crate that both Lightbeam and Cranelift could use?

**Comment by sunfishcode** [comment]
This is ready to go; it just needs a [few TODO comments](https://github.com/bytecodealliance/wasmtime/pull/1232#discussion_r406535349) and a rebase to fix some merge conflicts.

**Comment by sunfishcode** [comment]
Thanks!
